docs/docs/aztec/concepts/smart_contracts/functions/inner_workings.md

```diff
@@ -94,3 +98,269 @@ This:
+
+When a function is annotated with `#[aztec(noinitcheck)]`:
+
+- The Aztec macro processor skips the [insertion of the initialization check](#initializer-functions-aztecinitializer) for this specific function
+- The function can be called at any time, even if the contract hasn't been initialized yet
+
+## `Internal` functions #[aztec(internal)]
+
+This macro inserts a check at the beginning of the function to ensure that the caller is the contract itself. This is done by adding the following assertion:
+
+```rust
+assert(context.msg_sender() == context.this_address(), "Function can only be called internally");
+```
+
+## Custom notes #[aztec(note)]
+
+The `#[aztec(note)]` attribute is used to define custom note types in Aztec contracts. Learn more about notes [here](../../../concepts/storage/index.md).
+
+When a struct is annotated with `#[aztec(note)]`, the Aztec macro applies a series of transformations and generates implementations to turn it into a note that can be used in contracts to store private data.
+
+1. **NoteInterface Implementation**: The macro automatically implements most methods of the `NoteInterface` trait for the annotated struct. This includes:
+
+   - `serialize_content` and `deserialize_content`
+   - `get_header` and `set_header`
+   - `get_note_type_id`
+   - `compute_note_content_hash`
+   - `to_be_bytes`
+   - A `properties` method in the note's implementation
+
+2. **Automatic Header Field**: If the struct doesn't already have a `header` field of type `NoteHeader`, one is automatically created
+
+3. **Note Type ID Generation**: A unique `note_type_id` is automatically computed for the note type using a Keccak hash of the struct name
+
+4. **Serialization and Deserialization**: Methods for converting the note to and from a series of `Field` elements are generated, assuming each field can be converted to/from a `Field`
+
+5. **Property Metadata**: A separate struct is generated to describe the note's fields, which is used for efficient retrieval of note data
+
+6. **Export Information**: The note type and its ID are automatically exported
+
+
+### Before expansion
+
+Here is how you could define a custom note:
+
+```rust
+#[aztec(note)]
+struct CustomNote {
+    data: Field,
+    owner: Address,
+}
+```
+
+### After expansaion
+
+```rust
+impl CustomNote {
+    fn serialize_content(self: CustomNote) -> [Field; NOTE_SERIALIZED_LEN] {
+        [self.data, self.owner.to_field()]
+    }
+
+    fn deserialize_content(serialized_note: [Field; NOTE_SERIALIZED_LEN]) -> Self {
+        CustomNote {
+            data: serialized_note[0] as Field,
+            owner: Address::from_field(serialized_note[1]),
+            header: NoteHeader::empty()
+        }
+    }
+
+    fn get_note_type_id() -> Field {
+        // Automatically generated unique ID based on Keccak hash of the struct name
```

Comment:
Is the struct name the only input to the hash to generate the id? is there any possibility / concern with note type id collisions?

Reply:
currently yes it looks like this, but it's going to be nuked and i assume a better note type ID fn will exist. will have to keep an eye on this https://github.com/AztecProtocol/aztec-packages/blob/8b3dfe9dabc19e24f759b1a6c8ed14e2d9874149/noir/noir-repo/aztec_macros/src/transforms/note_interface.rs#L687